OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs

```diff
@@ -0,0 +1,74 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+	{
+		[Desc("The upgrades to grant when deploying and revoke when undeploying.")]
+		public readonly string[] Upgrades = { };
+
+		public object Create(ActorInitializer init) { return new DeployToUpgrade(init.Self, this); }
+	}
+
+	public class DeployToUpgrade : IResolveOrder, IIssueOrder
+	{
+		readonly DeployToUpgradeInfo info;
+		readonly UpgradeManager manager;
+
+		bool isUpgraded;
+
+		public DeployToUpgrade(Actor self, DeployToUpgradeInfo info)
+		{
+			this.info = info;
+			manager = self.Trait<UpgradeManager>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get { yield return new DeployOrderTargeter("DeployUpgrade", 5, () => true); }
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID == "DeployUpgrade")
+				return new Order(order.OrderID, self, queued);
+
+			return null;
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			// DeployTransform is necessary to use the DeployKey (F)
+			if (order.OrderString != "DeployUpgrade" && order.OrderString != "DeployTransform")
+				return;
+
+			if (isUpgraded)
+			{
+				foreach (var up in info.Upgrades)
+					manager.RevokeUpgrade(self, up, this);
+
+				isUpgraded = false;
+			}
+			else
+			{
+				foreach (var up in info.Upgrades)
+					manager.GrantUpgrade(self, up, this);
+
+				isUpgraded = true;
+			}
+		}
```

Comment:
also cant you omit the isUpgraded = false and true there, take the braces and set isUpgraded = !isUpgraded.
aka
```
			if (isUpgraded)
				foreach (var up in info.Upgrades)
					manager.RevokeUpgrade(self, up, this);
			else
				foreach (var up in info.Upgrades)
					manager.GrantUpgrade(self, up, this);

			isUpgraded = !isUpgraded ;
```